gtda/diagrams/features.py

```diff
@@ -701,3 +707,150 @@ def transform(self, X, y=None):
+
+    def __init__(self, order=1., n_bins=100, n_jobs=None):
+        self.order = order
+        self.n_bins = n_bins
+        self.n_jobs = n_jobs
+
+    def fit(self, X, y=None):
+        """Store all observed homology dimensions in
+        :attr:`homology_dimensions_` and, for each dimension separately,
+        store evenly sample filtration parameter values in :attr:`samplings_`.
+        Then, return the estimator.
+
+        This method is here to implement the usual scikit-learn API and hence
+        work in pipelines.
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples, n_features, 3)
+            Input data. Array of persistence diagrams, each a collection of
+            triples [b, d, q] representing persistent topological features
+            through their birth (b), death (d) and homology dimension (q).
+
+        y : None
+            There is no need for a target in a transformer, yet the pipeline
+            API requires this parameter.
+
+        Returns
+        -------
+        self : object
+
+        """
+        X = check_diagram(X)
+        validate_params(self.get_params(), self._hyperparameters)
+
+        self.homology_dimensions_ = sorted(list(set(X[0, :, 2])))
+        self._n_dimensions = len(self.homology_dimensions_)
+
+        self._samplings, _ = _bin(X, metric='silhouette', n_bins=self.n_bins)
+        self.samplings_ = {dim: s.flatten()
+                           for dim, s in self._samplings.items()}
+
+        return self
+
+    def transform(self, X, y=None):
+        """Compute silhouettes of diagrams in X.
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples, n_features, 3)
+            Input data. Array of persistence diagrams, each a collection of
+            triples [b, d, q] representing persistent topological features
+            through their birth (b), death (d) and homology dimension (q).
+
+        y : None
+            There is no need for a target in a transformer, yet the pipeline
+            API requires this parameter.
+
+        Returns
+        -------
+        Xt : ndarray of shape (n_samples, n_homology_dimensions, n_bins)
+            silhouette: one curve (represented as a one-dimensional array)
+            per sample and per homology dimension seen
+            in :meth:`fit`. Index i along axis 1 corresponds to the i-th
+            homology dimension in :attr:`homology_dimensions_`.
+
+        """
+        check_is_fitted(self)
+        X = check_diagram(X)
+
+        Xt = (Parallel(n_jobs=self.n_jobs)
+              (delayed(silhouettes)(_subdiagrams(X, [dim], remove_dim=True)[s],
+                                    self._samplings[dim], order=self.order)
+              for dim in self.homology_dimensions_
+              for s in gen_even_slices(X.shape[0],
+                                       effective_n_jobs(self.n_jobs))))
+        Xt = np.stack(Xt, axis=1)
```

Comment:
I don't think this is right. See https://gist.github.com/ulupo/ce732dfc39281d10e25d85aa954d9150 where I show that the shapes depend on the number of jobs. I think you'll have to mimic the solution used e.g. in `BettiCurve`.

Reply:
Ulalala, you are right!